### PR TITLE
secboot: move to new version

### PR DESCRIPTION
**Pull request by xnox**
Cherrypick @mvo5 move to go latest secboot using go mod
Cherrypick @mvo5 testcase improvement to `TestResealKey`
Cherrypick @chrisccoulson port to new snapcore/secboot/efi
Cherrypick/rebase @xnox port to new snapcore/secboot/tpm2

### CODE REVIEW

**Comment by anonymouse64** [comment]
FYI spread for uc20 is broken right now so you will need to wait until we unbreak things later today before you see meaningful spread results here, if you like I can merge master into this PR for you when that's ready

File: secboot/encrypt_sb.go
Comment (mvo5) [inline secboot/encrypt_sb.go:52]: Has Chris added this already to the new API? We need to keep this for one of our customers unfortunately.

Comment (xnox) [reply secboot/encrypt_sb.go:52]: @mvo5 @chrisccoulson 

Please see the detailed commit description describing what was done in secboot.

https://github.com/snapcore/snapd/pull/10715/commits/dde591333ade7fd021fc2ad17519e1975f4476af

Please let me know if what was done in secboot is sufficient, or not.

**Comment by xnox** [comment]
@anonymouse64  This is now in a state that it builds code, builds unit tests, but then fails unit tests.

I think I'm not mocking enough of the right things. For example, some apis have changed from accepting file names, to accepting objects. And it means that one has to mock real files, for them to be read and parsed into data. But i'm not too sure how to correctly do that in go. So I need help to get unittests modified to actually pass.

**Comment by anonymouse64** [comment]
@xnox I can try to have a peek at the unit tests here after we have finished unbreaking the UC20 world in GCE

File: secboot/secboot_tpm.go
Comment (xnox) [inline secboot/secboot_tpm.go:59]: @anonymouse64 when @chrisccoulson did the port from sb.FOO to sb_efi.FOO he renamed the constants from sbFOO to sbefiFoo.

I'm now wondering if my sb.FOO to sb_tpm2.FOO port should follow suite and rename sbFOO things to sbtpm2FOO.

Is there any guidance around this? ie. should I rename sbConnectToDefaultTPM  to sbtpm2ConnectToDefaultTPM and so on?

File: secboot/secboot_tpm.go
Comment (xnox) [inline secboot/secboot_tpm.go:394]: @anonymouse64 I believe this API change is breaking unit tests, as they are not expecting an extra call to `ReadSealedKeyObject()` which is not mocked and throws real "cannot read file" instead of being mocked.

File: secboot/encrypt_sb.go
Comment (anonymouse64) [inline secboot/encrypt_sb.go:52]: this comment should be removed as well if we are no longer using our own heuristic and instead using the heuristic of upstream secboot

File: secboot/encrypt_sb.go
Comment (anonymouse64) [reply secboot/encrypt_sb.go:52]: Seems to be the same to me

File: secboot/secboot_sb_test.go
Comment (anonymouse64) [inline secboot/secboot_sb_test.go:41]:
```suggestion
	"github.com/canonical/go-tpm2/linux"
	. "gopkg.in/check.v1"

	sb "github.com/snapcore/secboot"
	sb_efi "github.com/snapcore/secboot/efi"
	sb_tpm2 "github.com/snapcore/secboot/tpm2"
```

(and then I think the imports need to be re-order alphabetically as well, but the grouping is stdlib, external, non-snapd repos, then snapd internal packages)

File: secboot/secboot_tpm.go
Comment (anonymouse64) [reply secboot/secboot_tpm.go:59]: the pattern you have here is plenty readable and fine to me, perhaps @pedronis has further opinions but I think this is fine

File: secboot/secboot_tpm.go
Comment (anonymouse64) [reply secboot/secboot_tpm.go:394]: probably what you need to do is add to this package:

```go
var sb_tpm2ReadSealedKeyObject = sb_tpm2.ReadSealedKeyObject
```

then here use `sb_tpm2ReadSealedKeyObject`

then in export_test.go add 

```go
func MockReadSealedKeyObject(func f(string) (obj, error)) (restore func()) {
    old := sb_tpm2ReadSealedKeyObject
    sb_tpm2ReadSealedKeyObject = f
    return func() {
        sb_tpm2ReadSealedKeyObject = old
    }
}
```

then all the unit tests which call ResealKeys will be updated to call MockReadSealedKeyObject() with some function which checks/tracks that the calls to sb_tpm2ReadSealedKeyObject are as expected.

Does that make sense?

File: vendor/vendor.json
Comment (anonymouse64) [inline vendor/vendor.json:9]: mmm FYI we are very close to switching over to go modules, so this may become obsolete by the time it comes to actually merge this however managing dependencies with go modules is much easier and "the right thing" should be done automatically if you turn on go modules.

**Comment by pedronis** [comment]
afawk there's need to be more work in secboot around KDF options before we can switch to master there

File: secboot/encrypt_sb.go
Comment (mvo5) [reply secboot/encrypt_sb.go:52]: Thanks, I think we need https://github.com/snapcore/secboot/pull/174 in seccboot and then use that. If we drop the KDFOption we will regress quite a bit in install time on some devices.

**Comment by xnox** [comment]
> afawk there's need to be more work in secboot around KDF options before we can switch to master there

well the kdf options are available if one uses secboot/luks directly, rather than just secboot. Should I attempt to use that? Or do you want me to try to expose kdf options, the way they were exposed in the secboot fork, such that they are still available in new world order?

**Comment by pedronis** [comment]
> well the kdf options are available if one uses secboot/luks directly, rather than just secboot. Should I attempt to use that? Or do you want me to try to expose kdf options, the way they were exposed in the secboot fork, such that they are still available in new world order?

https://github.com/snapcore/secboot/pull/174 should help with this

File: secboot/encrypt_sb.go
Comment (mvo5) [reply secboot/encrypt_sb.go:52]: I update a PR for you in https://github.com/xnox/snapd/pull/1 - I can also directly merge in here but that felt a little bit impolite.

**Comment by bboozzoo** [comment]
This also needs to be updated to use go.mod now.

File: secboot/encrypt_sb.go
Comment (xnox) [reply secboot/encrypt_sb.go:52]: Ok secboot one landed now! great. Let me review what you did here, and we should be to land one thing or another.

**Comment by xnox** [comment]
```
Running vet
# github.com/canonical/go-efilib
../../../../pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210904134225-fd6cb98c6228/vars.go:82:16: undefined: os.ReadDir
```

Is this something that needs to be fixed in go-efilib? or like do I need to bump go modded version of go-efilib?!

**Comment by bboozzoo** [comment]
> ```
> Running vet
> # github.com/canonical/go-efilib
> ../../../../pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210904134225-fd6cb98c6228/vars.go:82:16: undefined: os.ReadDir
> ```
> 
> Is this something that needs to be fixed in go-efilib? or like do I need to bump go modded version of go-efilib?!

hmm pretty sure go vet should not be looking there at all

**Comment by xnox** [comment]
> > ```
> > Running vet
> > # github.com/canonical/go-efilib
> > ../../../../pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210904134225-fd6cb98c6228/vars.go:82:16: undefined: os.ReadDir
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this something that needs to be fixed in go-efilib? or like do I need to bump go modded version of go-efilib?!
> 
> hmm pretty sure go vet should not be looking there at all

I think this is fixed in go-efilib upstream, but it is pulled in indirectly, hence asking secboot to update the efilib https://github.com/snapcore/secboot/pull/175

**Comment by xnox** [comment]
@bboozzoo i'm failing to govendor bump things in https://github.com/snapcore/secboot/pull/175 and I fail to get 1.13 go to pass. But latest go passes? maybe we need to fix up go vet somehow on 1.13 to be less strict?

**Comment by bboozzoo** [comment]
> @bboozzoo i'm failing to govendor bump things in [snapcore/secboot#175](https://github.com/snapcore/secboot/pull/175) and I fail to get 1.13 go to pass. But latest go passes? maybe we need to fix up go vet somehow on 1.13 to be less strict?

I'll check out the branch locally and see what's going on there

**Comment by xnox** [comment]
> > > > ```
> > > > Running vet
> > > > # github.com/canonical/go-efilib
> > > > ../../../../pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210904134225-fd6cb98c6228/vars.go:82:16: undefined: os.ReadDir
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Is this something that needs to be fixed in go-efilib? or like do I need to bump go modded version of go-efilib?!
> > > 
> > > 
> > > hmm pretty sure go vet should not be looking there at all
> > 
> > 
> > I think this is fixed in go-efilib upstream, but it is pulled in indirectly, hence asking secboot to update the efilib [snapcore/secboot#175](https://github.com/snapcore/secboot/pull/175)
> > but also seems odd that go-efilib fails to build.
> 
> Yeah, I temporarily broke this when I pushed a change that introduced a dependency on go1.16, but nothing depends on the latest code at the moment - secboot currently depends on [canonical/go-efilib@7d3f2e2](https://github.com/canonical/go-efilib/commit/7d3f2e29397f8cf8b516710c609fde8bd6fe0550)

Hm, but we have

```
go.mod: github.com/canonical/go-efilib v0.0.0-20210904134225-fd6cb98c6228 // indirect
go.sum:github.com/canonical/go-efilib v0.0.0-20210904134225-fd6cb98c6228 h1:QMyqs+UL7qIALPR4Werst2XgXFIxmri46ZM/XlUjLnU=
go.sum:github.com/canonical/go-efilib v0.0.0-20210904134225-fd6cb98c6228/go.mod h1:9Sr9kd7IhQPYqaU5nut8Ky97/CtlhHDzQncQnrULgDM=
```

How do i downgrade to 7d3f2e29397f8cf8b516710c609fde8bd6fe0550 ? Not sure how to use `go mod` properly.

**Comment by xnox** [comment]
I think i figured it out. Let's see if it will work.

**Comment by bboozzoo** [comment]
> I think i figured it out. Let's see if it will work.

Interesting, I haven't seen any problems locally with the previous version either.

**Comment by xnox** [comment]
> > I think i figured it out. Let's see if it will work.
> 
> Interesting, I haven't seen any problems locally with the previous version either.

me neither, with golang 1.16, the test failed only with golang 1.13.

**Comment by xnox** [comment]
There are three warnings, from a passing check. I don't know if this matters:

```
run golangci-lint
  Running [/home/runner/golangci-lint-1.40.0-linux-amd64/golangci-lint run --out-format=github-actions --new-from-rev=master --path-prefix= --issues-exit-code=0] in [/home/runner/work/snapd/snapd/src/github.com/snapcore/snapd] ...
  Error: File is not `gci`-ed with -local github.com/snapcore/snapd (gci)
  Error: File is not `gci`-ed with -local github.com/snapcore/snapd (gci)
  Error: File is not `gci`-ed with -local github.com/snapcore/snapd (gci)
  
  golangci-lint found no issues
  Ran golangci-lint in 99380ms
```

File: secboot/secboot_sb_test.go
Comment (bboozzoo) [inline secboot/secboot_sb_test.go:43]: this should make the linter happy again:

```
diff --git a/secboot/secboot_sb_test.go b/secboot/secboot_sb_test.go
index 8388067634..6cbc50bb89 100644
--- a/secboot/secboot_sb_test.go
+++ b/secboot/secboot_sb_test.go
@@ -32,14 +32,12 @@ import (
        "os"
        "path/filepath"
 
-       . "gopkg.in/check.v1"
-
        "github.com/canonical/go-tpm2"
        "github.com/canonical/go-tpm2/linux"
-
        sb "github.com/snapcore/secboot"
        sb_efi "github.com/snapcore/secboot/efi"
        sb_tpm2 "github.com/snapcore/secboot/tpm2"
+       . "gopkg.in/check.v1"
 
        "github.com/snapcore/snapd/asserts"
        "github.com/snapcore/snapd/asserts/assertstest"
```

Comment (xnox) [reply secboot/secboot_sb_test.go:43]: but @anonymouse64 told me to do the reverse. i am confused.

File: secboot/secboot_sb_test.go
Comment (chrisccoulson) [inline secboot/secboot_sb_test.go:38]: Does this build ok? github.com/canonical/go-tpm2/linux only provides a single API - a transmission interface for the Linux TPM character device, which is used internally by secboot. I'd be surprised if this were required in snapd for testing.

File: secboot/secboot_sb_test.go
Comment (xnox) [inline secboot/secboot_sb_test.go:1111]: @chrisccoulson it's used here, inside a mock.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:38]: It is used inside a mock https://github.com/snapcore/snapd/pull/10715/files/81ee7783f70761a6819fb337778997710f188814#r703699848

File: secboot/secboot_sb_test.go
Comment (anonymouse64) [reply secboot/secboot_sb_test.go:43]: I can't seem to find what I told you, but I think I probably said something to the effect of "separate imports with external coming first, then internal", and the issue you have is that you may have conflated my saying "external" with "external to canonical/snapcore org" when really I meant external as external to the snapd repo, so we have this as the desired pattern:
```go
import (
    "standard library from go like fmt"

    "external from the snapd repo like github.com/snapcore/secboot or github.com/canonical/efilib or gopkg.in/gocheck.v1" 

    "internal to snapd repo like github.com/snapcore/snapd/osutil/..."
)
```

does that make sense ?

Comment (anonymouse64) [reply secboot/secboot_sb_test.go:43]: it's also entirely possible that I told you the inverse and I told you to put internal deps first and external second, but if I said that I was wrong, and I apologize for the confusion

File: secboot/secboot_sb_test.go
Comment (chrisccoulson) [reply secboot/secboot_sb_test.go:1111]: Ah, I didn't see that - thanks!

**Comment by anonymouse64** [comment]
there are now build failures for 32-bit platforms:

```
# github.com/canonical/go-efilib pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210708205505-7d3f2e29397f/vars.go:56:13: constant 3730735588 overflows int32
```

so it seems that we need an even newer version than 7d3f2e29397f. The newest version on main branch seems to not have the offending code, but it's unclear to me if the rest of secboot is compatible with the changes there as it seems to have changed its exported API? OpenEnhancedAuthenticatedVar seems to have disappeared at least...

**Review by anonymouse64 — COMMENTED** [review]
I think this PR is okay, one test nitpick to address, but there are still woes from the various efi/secboot dependencies it seems :-/

**Comment by xnox** [comment]
> there are now build failures for 32-bit platforms:
> 
> ```
> # github.com/canonical/go-efilib pkg/mod/github.com/canonical/go-efilib@v0.0.0-20210708205505-7d3f2e29397f/vars.go:56:13: constant 3730735588 overflows int32
> ```
> 
> so it seems that we need an even newer version than 7d3f2e29397f. The newest version on main branch seems to not have the offending code, but it's unclear to me if the rest of secboot is compatible with the changes there as it seems to have changed its exported API? OpenEnhancedAuthenticatedVar seems to have disappeared at least...

So this issue is because linux is written in C and constants in C do not care about being signed or unsigned.

But when golang processes constants into golang bindings it makes constants signed, and verifies that they fit into signed/unsigned ranges of syscalls. Which Linux itself does not care about. https://github.com/canonical/go-efilib/pull/3

I think golang itself should fix its bindings, or we need to convert the value from the syscall into bytes array, and convert constant into bytes array and then compare the two as bytes arrays. Rather than same length signed/unsigned numbers.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:1038]: Unlike auth file, the keyfile/keyfile2 are just strings rather than files that can be opened and read into SealedKeyObject struct. When I tried to create more full SealedKeyObjects, it was not allowed, as none of the fields are exported. I guess we need to stop having keyfile/keyfile2 as just strings, and have them as actual files, write them out to temp directory, to read them and process them into SealedKeyObjects?

Or is there some other API i can use to generate and created SealedKeyObject?

Comment (xnox) [reply secboot/secboot_sb_test.go:1036]: The error is already on the tc, and its being mocked and tested.

I can improve mockSealedKeyObjects to be as goot ad mockTPMPolicyAuthKey which is not on tc either.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:1049]: which is not done for mockTPMPolicyAuthKey, it's not in tc struct. So I don't see a need to add it.

Note that none of tc sub-cases today can change PolicyAuthKey, or have variable number of keyfiles.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:1040]: It would also fail at runtime trying to return non-existent array index... which should be just as obvious, no?!

File: secboot/secboot_sb_test.go
Comment (xnox) [inline secboot/secboot_sb_test.go:978]: @anonymouse64 please re-review this bit. I copied a sample test keyfileobject from snapcore/secboot and now i read enough of them to satisfy all calls of keyFiles args of myParams. This way all test cases match all existing requirements. Even if in the future keyFiles is changed.

File: secboot/secboot_sb_test.go
Comment (xnox) [inline secboot/secboot_sb_test.go:977]: @anonymouse64 please re-review this bit. I copied a sample test keyfileobject from snapcore/secboot and now i read enough of them to satisfy all calls of keyFiles args of myParams. This way all test cases match all existing requirements. Even if in the future keyFiles is changed.

Comment (anonymouse64) [reply secboot/secboot_sb_test.go:977]: looks good to me, thanks for this

**Review by anonymouse64 — APPROVED** [review]
the tests look good to me now, and it appears progress is being made on go-efilib so we probably just need to wait for that to be sorted out before we can again update the deps here and then this is one step closer to landing.

to be clear, we need 2 +1's from snapd maintainers, and for some things we like to have 3 +1s, I think this PR is one of the cases where we want 3 +1s

**Comment by xnox** [comment]
So we need https://github.com/snapcore/secboot/pull/176 merged, and then bump to that secboot with new go-efilib which should make things build & pass on 32bit.

**Comment by xnox** [comment]
@pedronis KDFoptions are back in place, can the blocked label be removed now?

File: secboot/secboot_sb_test.go
Comment (mvo5) [inline secboot/secboot_sb_test.go:973]: How is this keyfile generatd?

Comment (xnox) [reply secboot/secboot_sb_test.go:973]: It is a copy of https://github.com/snapcore/secboot/blob/master/internal/compattest/testdata/v1/key (i.e. any valid looking key that can be read into the sealed key object struct)

**Comment by xnox** [comment]
@anonymouse64 32bit stuff is now passing. There are a few required things failing still. Not sure if I just need to keep retrying / merging the development tip, or if some actions are needed from me. Did not investigate closely, but at the moment it seems unrelated to the changes I am making.

**Comment by anonymouse64** [comment]
@xnox I had a look at the failures in the logs and I agree, none look related. I'll try to have another look at this today

**Comment by xnox** [comment]
just great, merged tip and even more things are failing now than before.

**Comment by anonymouse64** [comment]
@xnox sadly it seems the store is not having a good day today, I will re-try your tests this evening when things have hopefully calmed down a bit, but the previous run before you merged master looked good to me, so I think it's unlikely anything is really broken here

**Comment by anonymouse64** [comment]
still lgtm, I restarted the tests, maybe the store is in a happier place now and some tests pass :crossed_fingers:

**Review by anonymouse64 — APPROVED** [review]
lgtm

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:973]: @mvo5 ping

**Comment by xnox** [comment]
@chrisccoulson is this ok from your point of view? or do you have review comments that i need to fix?

File: secboot/secboot_sb_test.go
Comment (mvo5) [inline secboot/secboot_sb_test.go:970]: Actually, I'm not sure why this change is needed and what exactly it's doing - if I change to the original code (or git revert d52049308d3122b89ed25804459123a94abae703):
```
diff --git a/secboot/secboot_sb_test.go b/secboot/secboot_sb_test.go
index 4992c039de..6cbc50bb89 100644
--- a/secboot/secboot_sb_test.go
+++ b/secboot/secboot_sb_test.go
@@ -967,15 +967,6 @@ func (s *secbootSuite) TestResealKey(c *C) {
                        TPMPolicyAuthKeyFile: mockTPMPolicyAuthKeyFile,
                }
 
-               numMockSealedKeyObjects := len(myParams.KeyFiles)
-               mockSealedKeyObjects := make([]*sb_tpm2.SealedKeyObject, 0, numMockSealedKeyObjects)
-               for range myParams.KeyFiles {
-                       mockSealedKeyFile := filepath.Join("test-data", "keyfile")
-                       mockSealedKeyObject, err := sb_tpm2.ReadSealedKeyObject(mockSealedKeyFile)
-                       c.Assert(err, IsNil)
-                       mockSealedKeyObjects = append(mockSealedKeyObjects, mockSealedKeyObject)
-               }
-
                sequences := []*sb_efi.ImageLoadEvent{
                        {
                                Source: sb_efi.Firmware,
@@ -1042,6 +1033,7 @@ func (s *secbootSuite) TestResealKey(c *C) {
 
                // mock ReadSealedKeyObject
                readSealedKeyObjectCalls := 0
+               mockSealedKeyObjects := []*sb_tpm2.SealedKeyObject{{}, {}}
                restore = secboot.MockSbReadSealedKeyObject(func(keyfile string) (*sb_tpm2.SealedKeyObject, error) {
                        readSealedKeyObjectCalls++
                        c.Assert(keyfile, Equals, myParams.KeyFiles[readSealedKeyObjectCalls-1])
```
afaict this test is still fine. But maybe/probably I'm missing something?

**Comment by mvo5** [comment]
This looks very nice, thanks for your work on this! Two small questions inline but otherwise good to go. It would be lovely if @chrisccoulson could have a quick look as well if possible though.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:970]: @mvo so yes, i had a version that did `mockSealedKeyObjects := []*sb_tpm2.SealedKeyObject{{}, {}}` which @anonymouse64 rejected.

Please make up your mind if you want {} empty mocks, or realistic mocks, which are not needed here at all.

Comment (xnox) [reply secboot/secboot_sb_test.go:970]: See https://github.com/snapcore/snapd/pull/10715#pullrequestreview-748340551

Comment (mvo5) [reply secboot/secboot_sb_test.go:970]: Sorry, I did miss the backstory here, I think it's fine then, what Ian says makes sense (but it's also very heavily mocked so I would be ok with something simpler) anyway, let's not bike-shed this here, I will make my comment "resolved".

File: secboot/secboot_sb_test.go
Comment (chrisccoulson) [inline secboot/secboot_sb_test.go:1114]: This is ok, but I think it's probably a bug in go-tpm2 that linux.OpenDevice("/dev/null") opens arbitrary devices and doesn't return an error for non-TPM devices. As tpm2.NewTPMContext no longer returns an error, it's possible to call it with nil, although I think Close() would crash later on if the tests call that. It might be better in future just to call tpm2.NewTPMContext with a stub that implements the tpm2.TCTI interface (https://pkg.go.dev/github.com/canonical/go-tpm2#TCTI)

**Review by chrisccoulson — APPROVED** [review]
This looks ok to me - I did leave one comment, but it's ok for now.

File: secboot/secboot_sb_test.go
Comment (xnox) [reply secboot/secboot_sb_test.go:1114]: It's bug for bug, given that in the past tpm2.OpenTPMDevice didn't return an error either, and probably should have too been mocked to implement a stub and an interface.

**Review by mvo5 — APPROVED** [review]

**Pull request merged**
